src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs

```diff
@@ -551,10 +565,9 @@ public virtual ODataResource CreateResource(SelectExpandNode selectExpandNode, R
             HashSet<string> declaredPropertyNameSet = new HashSet<string>(resource.Properties.Select(p => p.Name));
             List<ODataProperty> dynamicProperties = new List<ODataProperty>();
             IEnumerable<KeyValuePair<string, object>> dynamicPropertiesToSelect =
-                dynamicPropertyDictionary.Where(
-                    x =>
-                        !selectExpandNode.SelectedDynamicProperties.Any() ||
-                        selectExpandNode.SelectedDynamicProperties.Contains(x.Key));
+                dynamicPropertyDictionary.Where(x =>
+                selectExpandNode.SelectedDynamicProperties == null // if SelectedDynamicProperties == null, means SelectAllDynamicProperties == true
```

Comment:
>// if SelectedDynamicProperties == null, means SelectAllDynamicProperties == true [](start = 67, length = 81)

then why not just check SelectAllDynamicProperties == true?

Reply:
We have the following scenarios:

1) SelectAllDynamicProperties == true && SelectedDynamicProperties == null

2 ) SelectAllDynamicProperties == false && ( SelectedDynamicProperties == null || SelectedDynamicProperties != null)

Line 527 tells us that the only (SelectAllDynamicProperties == true || SelectedDynamicProperties != null) can pass.

Therefore, at Line 568, we can ignore the "SelectAllDynamicProperties == true", because in this case, `SelectedDynamicProperties` should equal to null.

In other word, we can only test `SelectedDynamicProperties != null`, because in this case, "SelectAllDynamicProperties` == false.

Let me update the comments in the codes.

---
In reply to: [354994953](https://github.com/OData/WebApi/pull/1941#discussion_r354994953) [](ancestors = 354994953)